### PR TITLE
fix_: a message might match more than one filter

### PR DESCRIPTION
**Pull request by richard-ramos**
While attempting to identify an issue with missing requests that @chaitanyaprem run into, I noticed that it is possible that when iterating the messages, to have two filters that match a message. The code as it was before could swallow the value of `matched` for a succesfully match, hence the message would get removed from the `storeMsgIDs` map

I also added some tests and populated some empty variables from ReceivedMessage, that even tho we seem to not use them, it does not make sense to leave them empty

### CODE REVIEW

**Review by chaitanyaprem — APPROVED** [review]
LGTM

Thanks!

**Review by kaichaosun — APPROVED** [review]

File: wakuv2/common/filter_test.go [hunk 1]
```diff
@@ -0,0 +1,366 @@
...
+		topic[0] = 0x01
+
+		f.ContentTopics[BytesToTopic(topic)] = struct{}{}
+	}
+
+	key, err := crypto.GenerateKey()
+	if err != nil {
+		t.Fatalf("generateFilter 1 failed")
+		return nil, err
+	}
```
Comment (igor-sirotin) [inline wakuv2/common/filter_test.go]: y not `require`? Is it for the error message? 🤔 
And I see it in more places below, not just here.
```suggestion
    require.NoError(t, err)
```

File: wakuv2/common/filter.go [hunk 2]
```diff
@@ -235,23 +235,24 @@ func (fs *Filters) NotifyWatchers(recvMessage *ReceivedMessage) bool {
 	}
 
 	for _, watcher := range candidates {
-		matched = true
+		// Messages are decrypted successfully only once
 		if decodedMsg == nil {
 			decodedMsg = recvMessage.Open(watcher)
 			if decodedMsg == nil {
 				log.Debug("processing message: failed to open", "message", recvMessage.Hash().Hex(), "filter", watcher.id)
+				continue
```
Comment (igor-sirotin) [inline wakuv2/common/filter.go:243]: hmm, Isn't (in theory) a `contentTopic+pubsubTopic` collision is possible? Then we should try to decode the message with a different `watcher`, because it might have different keys 🤔

**Comment by igor-sirotin** [comment]
And thanks for the test ❤️

File: wakuv2/common/filter.go (hunk 2 above)
Comment (chaitanyaprem) [reply wakuv2/common/filter.go:243]: what do you mean by collision here?

it is not expected for 2 different cases to use same `contentTopic+pubsubTopic` , because it would just lead to un-necessary receiving of messages and ultimately getting dropped in application.

Comment (igor-sirotin) [reply wakuv2/common/filter.go:243]: @chaitanyaprem  I mean, there's a reason why we get a list of filters here, right? 🙂 
https://github.com/status-im/status-go/blob/2a91fba1fad6427f233a448a6d9b4af07b7ee182/wakuv2/common/filter.go#L182

There might be more than one filter. And they might have different keys to decode the received message (I guess).

Comment (chaitanyaprem) [reply wakuv2/common/filter.go:243]: ah...you mean more than 1 filter per `contentTopic+pubsubTopic`. 

but why would there be more than 1 filter for same set...that doesn't seem right. there can be more than 1 watchers per filter though that want the messages received on it.

Comment (igor-sirotin) [reply wakuv2/common/filter.go:243]: > but why would there be more than 1 filter for same set...that doesn't seem right

That's because there might be collisions. `contentTopic` is calculated as 4 last bytes of hash of the actual topic. So it is very possible to have different topics sharing same `contentTopic`.

https://github.com/status-im/status-go/blob/063756b4ed401928044baa44f9c7f62b4152b525/wakuv2/common/topic.go#L55-L64

If we would want to prevent it, we would have to keep the whole topic. But then we'll loose privacy.

Comment (chaitanyaprem) [reply wakuv2/common/filter.go:243]: ah right, Thanks for the clarification!

File: wakuv2/common/filter_test.go (hunk 1 above)
Comment (richard-ramos) [reply wakuv2/common/filter_test.go]: Because of copy-pasting from another test :)  
Will fix it.

File: wakuv2/common/filter.go (hunk 2 above)
Comment (richard-ramos) [reply wakuv2/common/filter.go:243]: > @igor-sirotin: hmm, Isn't (in theory) a contentTopic+pubsubTopic collision is possible? Then we should try to decode the message with a different watcher, because it might have different keys.

Is this question referring to the `// Messages are decrypted successfully only once` comment or to the usage of the `continue` in this piece of code?

If it is the former, encrypted messages can be decrypted with only a single key. I havent reviewed the code throughly but assuming we can have multiple watchers, as long as they have the same key, they can reuse the `decodedMsg` that will contain the decrypted message (Decrypting is considered an expensive operation so it's better to do it only once if you get succesful result) . Later `watcher.MatchMessage` will determine if the decrypted message can be read by the filter by comparing either the public key of the message or the symmetric key, since comparing bytes is cheap, and that way all the watchers that share the same key and are subscribed to same pubsub topic / content topic will receive the message

```go
	for _, watcher := range candidates {
		// Messages are decrypted successfully only once  <--------------
		if decodedMsg == nil {
			decodedMsg = recvMessage.Open(watcher)
			if decodedMsg == nil {
				log.Debug("processing message: failed to open", "message", recvMessage.Hash().Hex(), "filter", watcher.id)
				continue
			}
		}

		if watcher.MatchMessage(decodedMsg) {   <------------------------------
```

If the question refers to the usage of `continue`, do note that we are running a loop that iterates over candidate watchers. If it couldn't decrypt with the watcher from the current iteration, then it will try on the next iteration, etc.

Comment (igor-sirotin) [reply wakuv2/common/filter.go:243]: > _encrypted messages can be decrypted with only a single key._

Indeed, didn't think of this. 🤦  
It all makes sense now, thanks!

**Review by igor-sirotin — APPROVED** [review]
